File: test/integration.test.ts
import { spawn, ChildProcess } from 'child_process';
import { Studio } from '../src/studio';

// Helper function to send MCP requests to the server
function sendMcpRequest(
  commandArgs: string[],
  request: object,
  timeout: number = 5000
): Promise<any> {
  return new Promise((resolve, reject) => {
    const child = spawn('node', ['dist/cli.js', ...commandArgs], {
      stdio: ['pipe', 'pipe', 'pipe']
    });

    let output = '';
    let errorOutput = '';
    let hasResolved = false;

    const timeoutId = setTimeout(() => {
      if (!hasResolved) {
        hasResolved = true;
        child.kill();
        reject(new Error(`Request timed out after ${timeout}ms`));
      }
    }, timeout);

    child.stdout?.on('data', (data) => {
      output += data.toString();

      // Try to parse JSON response(s)
      const lines = output.split('\n').filter(line => line.trim());
      if (lines.length > 0) {
        try {
          const response = JSON.parse(lines[lines.length - 1]);
          if (!hasResolved) {
            hasResolved = true;
            clearTimeout(timeoutId);
            child.kill();
            resolve(response);
          }
        } catch (e) {
          // Not valid JSON yet, continue waiting
        }
      }
    });

    child.stderr?.on('data', (data) => {
      errorOutput += data.toString();
    });

    child.on('error', (error) => {
      if (!hasResolved) {
        hasResolved = true;
        clearTimeout(timeoutId);
        reject(new Error(`Server failed to start: ${error.message}`));
      }
    });

    child.on('exit', (code) => {
      if (!hasResolved) {
        hasResolved = true;
        clearTimeout(timeoutId);
        if (code !== 0) {
          reject(new Error(`Server exited with code ${code}: ${errorOutput}`));
        } else {
          reject(new Error('Server exited without sending response'));
        }
      }
    });

    // Send the request
    const requestJson = JSON.stringify(request);
    child.stdin?.write(requestJson + '\n');
    child.stdin?.end();
  });
}

describe('Studio MCP Server Integration', () => {
  beforeAll(async () => {
    // Make sure the project is built
    const { execSync } = require('child_process');
    try {
      execSync('npm run build', { stdio: 'inherit' });
    } catch (error) {
      console.error('Failed to build project:', error);
      throw error;
    }
  });

  describe('Basic MCP Protocol', () => {
    it('responds to ping requests', async () => {
      const request = {
        jsonrpc: '2.0',
        id: '1',
        method: 'ping'
      };

      const response = await sendMcpRequest(['echo', 'hello'], request);

      expect(response.jsonrpc).toBe('2.0');
      expect(response.id).toBe('1');
      expect(response.result).toEqual({});
    });

    it('responds to initialize requests', async () => {
      const request = {
        jsonrpc: '2.0',
        id: '1',
        method: 'initialize',
        params: {
          protocolVersion: '2024-11-05',
          capabilities: {},
          clientInfo: {
            name: 'test-client',
            version: '1.0.0'
          }
        }
      };

      const response = await sendMcpRequest(['echo', 'hello'], request);

      expect(response.jsonrpc).toBe('2.0');
      expect(response.id).toBe('1');
      expect(response.result).toHaveProperty('protocolVersion');
      expect(response.result).toHaveProperty('capabilities');
      expect(response.result).toHaveProperty('serverInfo');
      expect(response.result.serverInfo.name).toBe('studio-mcp');
    });
  });

  describe('Tools functionality', () => {
    describe('with simple echo command', () => {
      it('lists available tools', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '2',
          method: 'tools/list'
        };

        const response = await sendMcpRequest(['echo', 'hello', '[args...]'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('2');
        expect(response.result.tools).toBeInstanceOf(Array);
        expect(response.result.tools).toHaveLength(1);

        const tool = response.result.tools[0];
        expect(tool.name).toBe('echo');
        expect(tool.description).toBe('Run the shell command `echo hello [args...]`');
        expect(tool.inputSchema.type).toBe('object');
        expect(tool.inputSchema.properties).toHaveProperty('args');
      });

      it('executes simple echo command', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '3',
          method: 'tools/call',
          params: {
            name: 'echo',
            arguments: {
              args: ['hello', 'world']
            }
          }
        };

        const response = await sendMcpRequest(['echo', 'hello', '[args...]'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('3');
        expect(response.result.content).toBeInstanceOf(Array);
        expect(response.result.content[0].type).toBe('text');
        expect(response.result.content[0].text).toBe('hello hello world');
        expect(response.result.isError).toBe(false);
      });
    });

    describe('with blueprinted echo command', () => {
      it('lists blueprinted tool with proper schema', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '4',
          method: 'tools/list'
        };

        const response = await sendMcpRequest(['echo', '{{text#the text to echo}}'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('4');

        const tool = response.result.tools[0];
        expect(tool.name).toBe('echo');
        expect(tool.description).toBe('Run the shell command `echo {{text}}`');

        const schema = tool.inputSchema;
        expect(schema.properties).toHaveProperty('text');
        expect(schema.properties.text.type).toBe('string');
        expect(schema.properties.text.description).toBe('the text to echo');
        expect(schema.properties).not.toHaveProperty('args');
      });

      it('executes blueprinted command', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '5',
          method: 'tools/call',
          params: {
            name: 'echo',
            arguments: {
              text: 'Hello Blueprint!'
            }
          }
        };

        const response = await sendMcpRequest(['echo', '{{text#the text to echo}}'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('5');
        expect(response.result.content[0].type).toBe('text');
        expect(response.result.content[0].text).toBe('Hello Blueprint!');
        expect(response.result.isError).toBe(false);
      });

      it('executes blueprinted command with additional args', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '6',
          method: 'tools/call',
          params: {
            name: 'echo',
            arguments: {
              text: 'Hello',
              args: ['World', 'from', 'args']
            }
          }
        };

        const response = await sendMcpRequest(['echo', '{{text#the text to echo}}', '[args...]'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('6');
        expect(response.result.content[0].type).toBe('text');
        expect(response.result.content[0].text).toBe('Hello World from args');
        expect(response.result.isError).toBe(false);
      });

      it('handles blueprint arguments with spaces', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '7',
          method: 'tools/call',
          params: {
            name: 'echo',
            arguments: {
              text: 'Hello World with spaces'
            }
          }
        };

        const response = await sendMcpRequest(['echo', '{{text#the text to echo}}'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('7');
        expect(response.result.content[0].type).toBe('text');
        expect(response.result.content[0].text).toBe('Hello World with spaces');
        expect(response.result.isError).toBe(false);
      });

      it('handles mixed blueprint with spaces', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '8',
          method: 'tools/call',
          params: {
            name: 'echo',
            arguments: {
              text: 'Hello World'
            }
          }
        };

        const response = await sendMcpRequest(['echo', 'simon says {{text#the text to echo}}'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('8');
        expect(response.result.content[0].type).toBe('text');
        expect(response.result.content[0].text).toBe('simon says Hello World');
        expect(response.result.isError).toBe(false);
      });

      it('handles blueprint definitions with spaces around hash', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '9',
          method: 'tools/call',
          params: {
            name: 'echo',
            arguments: {
              text: 'Hello World'
            }
          }
        };

        const response = await sendMcpRequest(['echo', '{{text # the text to echo}}'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('9');
        expect(response.result.content[0].type).toBe('text');
        expect(response.result.content[0].text).toBe('Hello World');
        expect(response.result.isError).toBe(false);
      });

      it('handles blueprints without descriptions', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '10',
          method: 'tools/list'
        };

        const response = await sendMcpRequest(['echo', '{{text}}'], request);

        const tool = response.result.tools[0];
        expect(tool.name).toBe('echo');
        expect(tool.description).toBe('Run the shell command `echo {{text}}`');

        const schema = tool.inputSchema;
        expect(schema.properties).toHaveProperty('text');
        expect(schema.properties.text.type).toBe('string');
        expect(schema.properties.text.description).toBeUndefined();
      });

      it('executes blueprints without descriptions', async () => {
        const request = {
          jsonrpc: '2.0',
          id: '11',
          method: 'tools/call',
          params: {
            name: 'echo',
            arguments: {
              text: 'Hello Blueprint!'
            }
          }
        };

        const response = await sendMcpRequest(['echo', '{{text}}'], request);

        expect(response.jsonrpc).toBe('2.0');
        expect(response.id).toBe('11');
        expect(response.result.content[0].type).toBe('text');
        expect(response.result.content[0].text).toBe('Hello Blueprint!');
        expect(response.result.isError).toBe(false);
      });
    });
  });

  describe('Error handling', () => {
    it('handles command errors gracefully', async () => {
      const request = {
        jsonrpc: '2.0',
        id: '12',
        method: 'tools/call',
        params: {
          name: 'false',
          arguments: {}
        }
      };

      const response = await sendMcpRequest(['false'], request);

      expect(response.jsonrpc).toBe('2.0');
      expect(response.id).toBe('12');
      expect(response.result.isError).toBe(true);
      expect(response.result.content).toBeInstanceOf(Array);
    });

    it('handles nonexistent tools', async () => {
      const request = {
        jsonrpc: '2.0',
        id: '13',
        method: 'tools/call',
        params: {
          name: 'nonexistent',
          arguments: {}
        }
      };

      const response = await sendMcpRequest(['echo', 'hello'], request);

      expect(response.jsonrpc).toBe('2.0');
      expect(response.id).toBe('13');
      expect(response).toHaveProperty('error');
    });
  });

  describe('Complex commands', () => {
    it('works with git commands', async () => {
      const request = {
        jsonrpc: '2.0',
        id: '14',
        method: 'tools/list'
      };

      const response = await sendMcpRequest(['git', 'status', '[args...]'], request);

      const tool = response.result.tools[0];
      expect(tool.name).toBe('git');
      expect(tool.description).toBe('Run the shell command `git status [args...]`');
    });

    it('works with multiple blueprint arguments', async () => {
      const request = {
        jsonrpc: '2.0',
        id: '15',
        method: 'tools/list'
      };

      const response = await sendMcpRequest([
        'rails', 'generate',
        '{{generator#Rails generator name}}',
        '{{name#Resource name}}'
      ], request);

      const tool = response.result.tools[0];
      expect(tool.name).toBe('rails');
      expect(tool.description).toBe('Run the shell command `rails generate {{generator}} {{name}}`');

      const schema = tool.inputSchema;
      expect(schema.properties).toHaveProperty('generator');
      expect(schema.properties).toHaveProperty('name');
      expect(schema.properties).not.toHaveProperty('args');
    });
  });
});

File: internal/studio/integration_test.go
package studio

import (
	"bufio"
	"context"
	"encoding/json"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// MCPRequest represents a JSON-RPC request
type MCPRequest struct {
	JSONRPC string      `json:"jsonrpc"`
	ID      string      `json:"id"`
	Method  string      `json:"method"`
	Params  interface{} `json:"params,omitempty"`
}

// MCPResponse represents a JSON-RPC response
type MCPResponse struct {
	JSONRPC string      `json:"jsonrpc"`
	ID      string      `json:"id"`
	Result  interface{} `json:"result,omitempty"`
	Error   interface{} `json:"error,omitempty"`
}

// InitializeParams represents initialize request parameters
type InitializeParams struct {
	ProtocolVersion string                 `json:"protocolVersion"`
	Capabilities    map[string]interface{} `json:"capabilities"`
	ClientInfo      map[string]interface{} `json:"clientInfo"`
}

// sendMCPRequest spawns the Go binary and sends an MCP request over stdio
func sendMCPRequest(t *testing.T, commandArgs []string, request MCPRequest, timeout time.Duration) MCPResponse {
	// Build the project first
	projectRoot, err := filepath.Abs("../..")
	require.NoError(t, err)

	buildCmd := exec.Command("go", "build", "-o", "studio-mcp", ".")
	buildCmd.Dir = projectRoot
	err = buildCmd.Run()
	require.NoError(t, err, "Failed to build project")

	// Prepare the command
	binaryPath := filepath.Join(projectRoot, "studio-mcp")
	args := append([]string{}, commandArgs...)
	cmd := exec.Command(binaryPath, args...)

	// Set up pipes
	stdin, err := cmd.StdinPipe()
	require.NoError(t, err)

	stdout, err := cmd.StdoutPipe()
	require.NoError(t, err)

	stderr, err := cmd.StderrPipe()
	require.NoError(t, err)

	// Start the process
	err = cmd.Start()
	require.NoError(t, err)

	defer func() {
		cmd.Process.Kill()
		cmd.Wait()
	}()

	// For tools/ methods, we need to initialize first
	needsInit := strings.HasPrefix(request.Method, "tools/")

	if needsInit {
		// Send initialize request first
		initRequest := MCPRequest{
			JSONRPC: "2.0",
			ID:      "init",
			Method:  "initialize",
			Params: InitializeParams{
				ProtocolVersion: "2024-11-05",
				Capabilities:    map[string]interface{}{},
				ClientInfo: map[string]interface{}{
					"name":    "test-client",
					"version": "1.0.0",
				},
			},
		}

		initJSON, err := json.Marshal(initRequest)
		require.NoError(t, err)

		_, err = stdin.Write(append(initJSON, '\n'))
		require.NoError(t, err)
	}

	// Send the actual request
	requestJSON, err := json.Marshal(request)
	require.NoError(t, err)

	_, err = stdin.Write(append(requestJSON, '\n'))
	require.NoError(t, err)
	stdin.Close()

	// Read response with timeout
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	responseData := make(chan []byte, 10) // Buffer multiple responses
	errorData := make(chan string, 1)

	go func() {
		scanner := bufio.NewScanner(stdout)
		for scanner.Scan() {
			line := scanner.Text()
			if strings.TrimSpace(line) != "" {
				responseData <- []byte(line)
			}
		}
		close(responseData)
	}()

	go func() {
		scanner := bufio.NewScanner(stderr)
		var errLines []string
		for scanner.Scan() {
			errLines = append(errLines, scanner.Text())
		}
		if len(errLines) > 0 {
			errorData <- strings.Join(errLines, "\n")
		}
	}()

	// Collect all responses and find the one matching our request ID
	var targetResponse MCPResponse
	found := false

	for {
		select {
		case data, ok := <-responseData:
			if !ok {
				// Channel closed, we're done
				if !found {
					t.Fatalf("Did not receive response for request ID %s", request.ID)
				}
				return targetResponse
			}

			var response MCPResponse
			err := json.Unmarshal(data, &response)
			require.NoError(t, err, "Failed to parse JSON response: %s", string(data))

			// Check if this is the response we're looking for
			if response.ID == request.ID {
				targetResponse = response
				found = true
			}

		case errMsg := <-errorData:
			t.Fatalf("Process error: %s", errMsg)
		case <-ctx.Done():
			t.Fatalf("Request timed out after %v", timeout)
		}
	}
}

func TestStudioMCPServerIntegration(t *testing.T) {
	timeout := 5 * time.Second

	t.Run("BasicMCPProtocol", func(t *testing.T) {
		t.Run("responds to ping requests", func(t *testing.T) {
			request := MCPRequest{
				JSONRPC: "2.0",
				ID:      "1",
				Method:  "ping",
			}

			response := sendMCPRequest(t, []string{"echo", "hello"}, request, timeout)

			assert.Equal(t, "2.0", response.JSONRPC)
			assert.Equal(t, "1", response.ID)
			assert.Equal(t, map[string]interface{}{}, response.Result)
		})

		t.Run("responds to initialize requests", func(t *testing.T) {
			request := MCPRequest{
				JSONRPC: "2.0",
				ID:      "1",
				Method:  "initialize",
				Params: InitializeParams{
					ProtocolVersion: "2024-11-05",
					Capabilities:    map[string]interface{}{},
					ClientInfo: map[string]interface{}{
						"name":    "test-client",
						"version": "1.0.0",
					},
				},
			}

			response := sendMCPRequest(t, []string{"echo", "hello"}, request, timeout)

			assert.Equal(t, "2.0", response.JSONRPC)
			assert.Equal(t, "1", response.ID)

			result, ok := response.Result.(map[string]interface{})
			require.True(t, ok, "Result should be an object")

			assert.Contains(t, result, "protocolVersion")
			assert.Contains(t, result, "capabilities")
			assert.Contains(t, result, "serverInfo")

			serverInfo, ok := result["serverInfo"].(map[string]interface{})
			require.True(t, ok, "serverInfo should be an object")
			assert.Equal(t, "studio-mcp", serverInfo["name"])
		})
	})

	t.Run("ToolsFunctionality", func(t *testing.T) {
		t.Run("with simple echo command", func(t *testing.T) {
			t.Run("lists available tools", func(t *testing.T) {
				request := MCPRequest{
					JSONRPC: "2.0",
					ID:      "2",
					Method:  "tools/list",
				}

				response := sendMCPRequest(t, []string{"echo", "hello", "[args...]"}, request, timeout)

				assert.Equal(t, "2.0", response.JSONRPC)
				assert.Equal(t, "2", response.ID)

				result, ok := response.Result.(map[string]interface{})
				require.True(t, ok)

				tools, ok := result["tools"].([]interface{})
				require.True(t, ok)
				assert.Len(t, tools, 1)

				tool, ok := tools[0].(map[string]interface{})
				require.True(t, ok)
				assert.Equal(t, "echo", tool["name"])
				assert.Equal(t, "Run the shell command `echo hello [args...]`", tool["description"])

				inputSchema, ok := tool["inputSchema"].(map[string]interface{})
				require.True(t, ok)
				assert.Equal(t, "object", inputSchema["type"])

				properties, ok := inputSchema["properties"].(map[string]interface{})
				require.True(t, ok)
				assert.Contains(t, properties, "args")
			})

			t.Run("executes simple echo command", func(t *testing.T) {
				request := MCPRequest{
					JSONRPC: "2.0",
					ID:      "3",
					Method:  "tools/call",
					Params: map[string]interface{}{
						"name": "echo",
						"arguments": map[string]interface{}{
							"args": []string{"hello", "world"},
						},
					},
				}

				response := sendMCPRequest(t, []string{"echo", "hello", "[args...]"}, request, timeout)

				assert.Equal(t, "2.0", response.JSONRPC)
				assert.Equal(t, "3", response.ID)

				result, ok := response.Result.(map[string]interface{})
				require.True(t, ok)

				content, ok := result["content"].([]interface{})
				require.True(t, ok)
				require.Len(t, content, 1)

				textContent, ok := content[0].(map[string]interface{})
				require.True(t, ok)
				assert.Equal(t, "text", textContent["type"])
				assert.Equal(t, "hello hello world", textContent["text"])

				// isError should be false or omitted (omitted means false)
				if isError, exists := result["isError"]; exists {
					assert.Equal(t, false, isError)
				}
			})
		})

		t.Run("with blueprinted echo command", func(t *testing.T) {
			t.Run("lists blueprinted tool with proper schema", func(t *testing.T) {
				request := MCPRequest{
					JSONRPC: "2.0",
					ID:      "4",
					Method:  "tools/list",
				}

				response := sendMCPRequest(t, []string{"echo", "{{text#the text to echo}}"}, request, timeout)

				assert.Equal(t, "2.0", response.JSONRPC)
				assert.Equal(t, "4", response.ID)

				result, ok := response.Result.(map[string]interface{})
				require.True(t, ok)

				tools, ok := result["tools"].([]interface{})
				require.True(t, ok)
				require.Len(t, tools, 1)

				tool, ok := tools[0].(map[string]interface{})
				require.True(t, ok)
				assert.Equal(t, "echo", tool["name"])
				assert.Equal(t, "Run the shell command `echo {{text}}`", tool["description"])

				inputSchema, ok := tool["inputSchema"].(map[string]interface{})
				require.True(t, ok)

				properties, ok := inputSchema["properties"].(map[string]interface{})
				require.True(t, ok)
				assert.Contains(t, properties, "text")
				assert.NotContains(t, properties, "args")

				textProp, ok := properties["text"].(map[string]interface{})
				require.True(t, ok)
				assert.Equal(t, "string", textProp["type"])
				assert.Equal(t, "the text to echo", textProp["description"])
			})

			t.Run("executes blueprinted command", func(t *testing.T) {
				request := MCPRequest{
					JSONRPC: "2.0",
					ID:      "5",
					Method:  "tools/call",
					Params: map[string]interface{}{
						"name": "echo",
						"arguments": map[string]interface{}{
							"text": "Hello Blueprint!",
						},
					},
				}

				response := sendMCPRequest(t, []string{"echo", "{{text#the text to echo}}"}, request, timeout)

				assert.Equal(t, "2.0", response.JSONRPC)
				assert.Equal(t, "5", response.ID)

				result, ok := response.Result.(map[string]interface{})
				require.True(t, ok)

				content, ok := result["content"].([]interface{})
				require.True(t, ok)
				require.Len(t, content, 1)

				textContent, ok := content[0].(map[string]interface{})
				require.True(t, ok)
				assert.Equal(t, "text", textContent["type"])
				assert.Equal(t, "Hello Blueprint!", textContent["text"])

				// isError should be false or omitted (omitted means false)
				if isError, exists := result["isError"]; exists {
					assert.Equal(t, false, isError)
				}
			})
		})
	})

	t.Run("ErrorHandling", func(t *testing.T) {
		t.Run("handles command errors gracefully", func(t *testing.T) {
			request := MCPRequest{
				JSONRPC: "2.0",
				ID:      "12",
				Method:  "tools/call",
				Params: map[string]interface{}{
					"name":      "false",
					"arguments": map[string]interface{}{},
				},
			}

			response := sendMCPRequest(t, []string{"false"}, request, timeout)

			assert.Equal(t, "2.0", response.JSONRPC)
			assert.Equal(t, "12", response.ID)

			result, ok := response.Result.(map[string]interface{})
			require.True(t, ok)
			assert.Equal(t, true, result["isError"])

			content, ok := result["content"].([]interface{})
			require.True(t, ok)
			require.Len(t, content, 1)
		})

		t.Run("handles nonexistent tools", func(t *testing.T) {
			request := MCPRequest{
				JSONRPC: "2.0",
				ID:      "13",
				Method:  "tools/call",
				Params: map[string]interface{}{
					"name":      "nonexistent",
					"arguments": map[string]interface{}{},
				},
			}

			response := sendMCPRequest(t, []string{"echo", "hello"}, request, timeout)

			assert.Equal(t, "2.0", response.JSONRPC)
			assert.Equal(t, "13", response.ID)
			assert.NotNil(t, response.Error)
		})
	})
}
